File: java/src/main/java/qiniu/uip/db/format/ipdb/Ipdb.java
package qiniu.uip.db.format.ipdb;

import qiniu.uip.db.IPFormatException;
import qiniu.uip.db.InvalidDatabaseException;
import qiniu.uip.db.IpInfo;
import qiniu.uip.db.VersionInfo;
import qiniu.uip.db.format.Query;
import qiniu.uip.db.format.QueryBuilder;
import qiniu.uip.util.Strings;

public final class Ipdb implements Query {

    private final Reader reader;

    public Ipdb(byte[] dat) throws InvalidDatabaseException {
        reader = new Reader(dat);
    }


    @Override
    public IpInfo query(byte[] ip) throws IPFormatException, InvalidDatabaseException {
        return reader.query(ip, true, false);
    }

    @Override
    public IpInfo query(long ab, long cd) throws IPFormatException, InvalidDatabaseException {
        return null;
    }

    @Override
    public void buildCache(byte[][] ipList) {
        reader.buildCache(ipList);
    }

    public VersionInfo version() {
        VersionInfo versionInfo = new VersionInfo();
        String version;
        if (Strings.isEmpty(reader.meta.version)) {
            version = "ipip.net-" + reader.meta.build;
        } else {
            version = reader.meta.version;
        }
        versionInfo.ipType = reader.meta.ip_version;
        versionInfo.count = reader.meta.node_count;
        versionInfo.build = reader.meta.build;
        versionInfo.version = version;
        versionInfo.languages = new String[]{"CN"};
        versionInfo.extraInfo = reader.meta.extra;
        return versionInfo;
    }

    public static class Builder implements QueryBuilder {
        public Builder() {
        }

        public Ipdb build(byte[] dat) throws InvalidDatabaseException {
            return new Ipdb(dat);
        }
    }
}

File: rust/src/formats/ipdb.rs
use std::io;
use crate::formats::{Query, QueryBuilder};
use crate::data::{VersionInfo, IpInfo};
mod reader;

struct IpdbBuilder {
}

impl QueryBuilder for IpdbBuilder {
    fn build(bytes: &[u8]) -> io::Result<Box<dyn Query>> {
        let reader = Ipdb::new(bytes)?;
        Ok(Box::new(reader))
    }
}

pub struct Ipdb {
    reader: reader::Reader,
}

impl Ipdb {
    pub fn new(data: &[u8]) -> io::Result<Ipdb> {
        let reader = reader::Reader::new(&data)?;
        Ok(Ipdb { reader })
    }
}

impl Query for Ipdb {
    fn query(&self, ip:&[u8]) -> io::Result<IpInfo>{
        self.reader.find(ip, "CN", true)
    }

    fn build_cache(&mut self, _ip_list:&Vec<String>) {
        println!("ipdb build_cache");
    }

    fn version(&self) -> VersionInfo {
        println!("ipdb get_version");
        VersionInfo {
            ip_type: self.reader.meta.ip_version,
            count: self.reader.meta.node_count,
            build: self.reader.meta.build,
            version: match &self.reader.meta.version {
                None => format!("ipip.net-{}", self.reader.meta.build),
                Some(s) => s.to_string(),
            },
            languages: vec!["CN".into()],
            extra_info: self.reader.meta.extra.clone(),
        }
    }
}